File: frontend/src/types/capabilities.ts
export interface UploadCapabilities {
  extensions: string[];
  mime_types: Record<string, string[]>;
  max_size_bytes: number;
  max_pdf_pages: number;
  max_files_per_selection: number;
}

export interface BatchCapabilities {
  limit_options: number[];
  default_limit: number;
  max_limit: number;
}

export interface PollingCapabilities {
  interval_ms: number;
  max_attempts: number;
}

export interface ImplementationCapabilities {
  document_parser: string;
  ocr_engine: string;
  llm_provider: string;
  vector_store: string;
}

export interface PipelineStageCapability {
  id: string;
  label: string;
  description: string;
}

export interface SimilarityCapabilities {
  default_threshold: number;
  minimum_threshold: number;
  maximum_threshold: number;
  default_limit: number;
  maximum_limit: number;
  high_band: number;
  medium_band: number;
}

export interface ApplicationCapabilities {
  upload: UploadCapabilities;
  batch: BatchCapabilities;
  polling: PollingCapabilities;
  implementation: ImplementationCapabilities;
  pipeline: PipelineStageCapability[];
  similarity: SimilarityCapabilities;
}

File: backend/app/schemas/capabilities.py
from pydantic import BaseModel, Field


class UploadCapabilities(BaseModel):
    extensions: list[str]
    mime_types: dict[str, list[str]]
    max_size_bytes: int = Field(ge=1)
    max_pdf_pages: int = Field(ge=1)
    max_files_per_selection: int = Field(ge=1)


class BatchCapabilities(BaseModel):
    limit_options: list[int]
    default_limit: int = Field(ge=1)
    max_limit: int = Field(ge=1)


class PollingCapabilities(BaseModel):
    interval_ms: int = Field(ge=1)
    max_attempts: int = Field(ge=1)


class ImplementationCapabilities(BaseModel):
    document_parser: str
    ocr_engine: str
    llm_provider: str
    vector_store: str


class PipelineStageCapability(BaseModel):
    id: str = Field(min_length=1)
    label: str = Field(min_length=1)
    description: str = Field(min_length=1)


class ApplicationCapabilities(BaseModel):
    upload: UploadCapabilities
    batch: BatchCapabilities
    polling: PollingCapabilities
    implementation: ImplementationCapabilities
    pipeline: list[PipelineStageCapability]
